In [8]:
%pip install scikit-learn
%pip install pandas
%pip install numpy
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score


Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [9]:
print(f"="*60)
print(f"Part 1: Fuel Consumption -> Horsepower Prediction")
print(f"="*60)

df1 = pd.read_csv('/home/veronica/桌面/Homework1/FuelEconomy.csv')
print(f"\n1.1 Load and Inspect Dataset")
print(f"Column names: {list(df1.columns)}")
print(f"Shape: {df1.shape}")
print(f"\nSummary Statistics:")
print(df1.describe())
print(f"\nMissing values:")
print(df1.isnull().sum())

# get X and y, X is fuel economy, y is horsepower
X1 = df1[['Fuel Economy (MPG)']].values
y1 = df1['Horse Power'].values

# 1.2 split 70% train 30% test
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.3, random_state=42)
print(f"\nTraining samples: {len(X1_train)}")
print(f"Testing samples: {len(X1_test)}")

# 1.3 train all the models

# basic linear regression
lr1 = LinearRegression()
lr1.fit(X1_train, y1_train)

# ridge with alpha=1.0, i just use default
ridge1 = Ridge(alpha=1.0)
ridge1.fit(X1_train, y1_train)

# lasso also alpha=1.0
lasso1 = Lasso(alpha=1.0)
lasso1.fit(X1_train, y1_train)

# polynomial degree 2
poly2_1 = PolynomialFeatures(degree=2)
X1_train_poly2 = poly2_1.fit_transform(X1_train)
X1_test_poly2 = poly2_1.transform(X1_test)

# polynomial degree 3
poly3_1 = PolynomialFeatures(degree=3)
X1_train_poly3 = poly3_1.fit_transform(X1_train)
X1_test_poly3 = poly3_1.transform(X1_test)

# poly2 + linear
lr_poly2_1 = LinearRegression()
lr_poly2_1.fit(X1_train_poly2, y1_train)

# poly3 + linear
lr_poly3_1 = LinearRegression()
lr_poly3_1.fit(X1_train_poly3, y1_train)

# poly2 + ridge
ridge_poly2_1 = Ridge(alpha=1.0)
ridge_poly2_1.fit(X1_train_poly2, y1_train)

# poly3 + ridge
ridge_poly3_1 = Ridge(alpha=1.0)
ridge_poly3_1.fit(X1_train_poly3, y1_train)

# poly2 + lasso
lasso_poly2_1 = Lasso(alpha=1.0)
lasso_poly2_1.fit(X1_train_poly2, y1_train)

# poly3 + lasso
lasso_poly3_1 = Lasso(alpha=1.0)
lasso_poly3_1.fit(X1_train_poly3, y1_train)

print(f"All models trained, alpha = 1.0")

# 1.4 evaluation - calculate mse mae r2 for each model
res1 = []

y1_train_pred=lr1.predict(X1_train)
y1_test_pred=lr1.predict(X1_test)
res1.append(['Linear Regression', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=ridge1.predict(X1_train)
y1_test_pred=ridge1.predict(X1_test)
res1.append(['Ridge Regression', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=lasso1.predict(X1_train)
y1_test_pred=lasso1.predict(X1_test)
res1.append(['Lasso Regression', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=lr_poly2_1.predict(X1_train_poly2)
y1_test_pred=lr_poly2_1.predict(X1_test_poly2)
res1.append(['Poly2 + Linear', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=lr_poly3_1.predict(X1_train_poly3)
y1_test_pred=lr_poly3_1.predict(X1_test_poly3)
res1.append(['Poly3 + Linear', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=ridge_poly2_1.predict(X1_train_poly2)
y1_test_pred=ridge_poly2_1.predict(X1_test_poly2)
res1.append(['Poly2 + Ridge', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=ridge_poly3_1.predict(X1_train_poly3)
y1_test_pred=ridge_poly3_1.predict(X1_test_poly3)
res1.append(['Poly3 + Ridge', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=lasso_poly2_1.predict(X1_train_poly2)
y1_test_pred=lasso_poly2_1.predict(X1_test_poly2)
res1.append(['Poly2 + Lasso', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

y1_train_pred=lasso_poly3_1.predict(X1_train_poly3)
y1_test_pred=lasso_poly3_1.predict(X1_test_poly3)
res1.append(['Poly3 + Lasso', mean_squared_error(y1_train, y1_train_pred), mean_absolute_error(y1_train, y1_train_pred), r2_score(y1_train, y1_train_pred), mean_squared_error(y1_test, y1_test_pred), mean_absolute_error(y1_test, y1_test_pred), r2_score(y1_test, y1_test_pred)])

# make a table to show results
df_res1 = pd.DataFrame(res1, columns=['Model', 'Train MSE', 'Train MAE', 'Train R2', 'Test MSE', 'Test MAE', 'Test R2'])
print(df_res1.to_string(index=False))

# 1.5 find best model by test r2
best_idx=0
best_r2=res1[0][6]
i=1
while i<len(res1):
    if res1[i][6]>best_r2:
        best_r2=res1[i][6]
        best_idx=i
    i+=1
print(f"Best model: {res1[best_idx][0]} with test R2={best_r2:.4f}")


Part 1: Fuel Consumption -> Horsepower Prediction

1.1 Load and Inspect Dataset
Column names: ['Horse Power', 'Fuel Economy (MPG)']
Shape: (100, 2)

Summary Statistics:
       Horse Power  Fuel Economy (MPG)
count   100.000000          100.000000
mean    213.676190           23.178501
std      62.061726            4.701666
min      50.000000           10.000000
25%     174.996514           20.439516
50%     218.928402           23.143192
75%     251.706476           26.089933
max     350.000000           35.000000

Missing values:
Horse Power           0
Fuel Economy (MPG)    0
dtype: int64

Training samples: 70
Testing samples: 30
All models trained, alpha = 1.0
            Model  Train MSE  Train MAE  Train R2   Test MSE  Test MAE  Test R2
Linear Regression 357.699180  16.061689  0.906320 318.561087 14.940628 0.912561
 Ridge Regression 357.700562  16.063735  0.906320 318.806477 14.948544 0.912494
 Lasso Regression 357.743457  16.073268  0.906309 319.982962 14.985418 0.912171
   Poly2

Poly3 + Linear model has the best test R2 (0.9126), but it is very close to the basic Linear Regression (0.9126 vs 0.9126). Actually they are almost the same.

The linear models (Linear, Ridge, Lasso) all perform very well with test R2 around 0.912-0.913. This means the relationship between Fuel Economy and Horsepower is mostly linear. Adding polynomial features does not improve the model much.

Ridge and Lasso have similar performance to Linear Regression because this is a simple dataset with only one feature. There is not much overfitting to regularize.

The best generalizing model is Poly3 + Linear, but simple Linear Regression works almost as well. I think linear relationship is enough for this data.

In [20]:
print(f"\n\n")
print(f"="*60)
print(f"Part 2: Weather -> Daily Electricity Consumption Prediction")
print(f"="*60)

df2 = pd.read_csv('/home/veronica/桌面/Homework1/electricity_consumption_based_weather_dataset.csv')

# clean the csv thats whymissing data
df2 = df2.dropna()
print(f'After dropna shape: {df2.shape}')

feature_cols_2 = ['AWND', 'PRCP', 'TMAX', 'TMIN']
X2 = df2[feature_cols_2].values
y2 = df2['daily_consumption'].values

X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.3, random_state=42)
print(f'Training: {len(X2_train)}, Testing: {len(X2_test)}')

lr2 = LinearRegression()
lr2.fit(X2_train, y2_train)
ridge2 = Ridge(alpha=1.0)
ridge2.fit(X2_train, y2_train)
lasso2 = Lasso(alpha=1.0)
lasso2.fit(X2_train, y2_train)

poly2_2 = PolynomialFeatures(degree=2)
X2_train_poly2 = poly2_2.fit_transform(X2_train)
X2_test_poly2 = poly2_2.transform(X2_test)
poly3_2 = PolynomialFeatures(degree=3)
X2_train_poly3 = poly3_2.fit_transform(X2_train)
X2_test_poly3 = poly3_2.transform(X2_test)

lr_poly2_2 = LinearRegression()
lr_poly2_2.fit(X2_train_poly2, y2_train)
lr_poly3_2 = LinearRegression()
lr_poly3_2.fit(X2_train_poly3, y2_train)
ridge_poly2_2 = Ridge(alpha=1.0)
ridge_poly2_2.fit(X2_train_poly2, y2_train)
ridge_poly3_2 = Ridge(alpha=1.0)
ridge_poly3_2.fit(X2_train_poly3, y2_train)
lasso_poly2_2 = Lasso(alpha=1.0)
lasso_poly2_2.fit(X2_train_poly2, y2_train)
lasso_poly3_2 = Lasso(alpha=1.0)
lasso_poly3_2.fit(X2_train_poly3, y2_train)

res2 = []
y2_train_pred=lr2.predict(X2_train)
y2_test_pred=lr2.predict(X2_test)
res2.append(['Linear Regression', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=ridge2.predict(X2_train)
y2_test_pred=ridge2.predict(X2_test)
res2.append(['Ridge Regression', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=lasso2.predict(X2_train)
y2_test_pred=lasso2.predict(X2_test)
res2.append(['Lasso Regression', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=lr_poly2_2.predict(X2_train_poly2)
y2_test_pred=lr_poly2_2.predict(X2_test_poly2)
res2.append(['Poly2 + Linear', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=lr_poly3_2.predict(X2_train_poly3)
y2_test_pred=lr_poly3_2.predict(X2_test_poly3)
res2.append(['Poly3 + Linear', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=ridge_poly2_2.predict(X2_train_poly2)
y2_test_pred=ridge_poly2_2.predict(X2_test_poly2)
res2.append(['Poly2 + Ridge', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=ridge_poly3_2.predict(X2_train_poly3)
y2_test_pred=ridge_poly3_2.predict(X2_test_poly3)
res2.append(['Poly3 + Ridge', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=lasso_poly2_2.predict(X2_train_poly2)
y2_test_pred=lasso_poly2_2.predict(X2_test_poly2)
res2.append(['Poly2 + Lasso', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

y2_train_pred=lasso_poly3_2.predict(X2_train_poly3)
y2_test_pred=lasso_poly3_2.predict(X2_test_poly3)
res2.append(['Poly3 + Lasso', mean_squared_error(y2_train, y2_train_pred), mean_absolute_error(y2_train, y2_train_pred), r2_score(y2_train, y2_train_pred), mean_squared_error(y2_test, y2_test_pred), mean_absolute_error(y2_test, y2_test_pred), r2_score(y2_test, y2_test_pred)])

df_res2 = pd.DataFrame(res2, columns=['Model', 'Train MSE', 'Train MAE', 'Train R2', 'Test MSE', 'Test MAE', 'Test R2'])
print(df_res2.to_string(index=False))

best_idx=0
best_r2=res2[0][6]
i=1
while i<len(res2):
    if res2[i][6]>best_r2:
        best_r2=res2[i][6]
        best_idx=i
    i+=1
print(f'Best model: {res2[best_idx][0]} with test R2={best_r2:.4f}')




Part 2: Weather -> Daily Electricity Consumption Prediction
After dropna shape: (1418, 6)
Training: 992, Testing: 426
            Model     Train MSE  Train MAE  Train R2      Test MSE   Test MAE  Test R2
Linear Regression 272403.396174 384.465016  0.276000 248125.785784 375.404537 0.299333
 Ridge Regression 272403.396254 384.465310  0.276000 248125.799085 375.404855 0.299333
 Lasso Regression 272404.552102 384.508177  0.275997 248122.648810 375.439837 0.299342
   Poly2 + Linear 264765.769932 379.648753  0.296300 255268.493954 379.039083 0.279163
   Poly3 + Linear 259249.534870 375.952901  0.310961 265623.657673 385.235167 0.249922
    Poly2 + Ridge 264766.578998 379.585166  0.296298 254955.169280 378.881442 0.280048
    Poly3 + Ridge 259307.759576 376.033550  0.310806 257659.254512 383.568232 0.272412
    Poly2 + Lasso 264881.385611 379.509069  0.295993 253247.451059 378.033190 0.284870
    Poly3 + Lasso 261405.827082 375.970982  0.305230 249726.476276 376.697061 0.294813
Best mode

/home/veronica/桌面/Homework1/.venv/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.230e+08, tolerance: 3.732e+04
  model = cd_fast.enet_coordinate_descent(
/home/veronica/桌面/Homework1/.venv/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.299e+08, tolerance: 3.732e+04
  model = cd_fast.enet_coordinate_descent(


All models have very low test R2 scores (around 0.25-0.30). The best model is Lasso Regression with test R2 = 0.2993.

The linear models (Linear, Ridge, Lasso) actually perform better on test data than the polynomial models. The polynomial models have higher train R2 but lower test R2, which means they are overfitting the training data.

The low R2 scores (around 0.30) mean that weather features can only explain about 30% of the variance in electricity consumption. This suggests that electricity consumption depends on many other factors not in this dataset, like day of week, holidays, or human behavior. Weather alone is not a good predictor.

Since linear models generalize better than polynomial models, the relationship between weather and electricity consumption seems to be mostly linear, but just very weak overall.



In [11]:
print(f"\n\n")
print(f"="*60)
print(f"Part 3: Power Plant Data -> PE Prediction")
print(f"="*60)

df3 = pd.read_csv('/home/veronica/桌面/Homework1/usina.csv')
print(f"\n3.1 Load and Inspect Dataset")
print(f"Column names: {list(df3.columns)}")
print(f"Shape: {df3.shape}")
print(f"\nSummary Statistics:")
print(df3.describe())
print(f"\nMissing values:")
print(df3.isnull().sum())
print(f"\nTarget (DV): PE (Electrical Energy Output)")
print(f"Features (IVs): AT (Ambient Temperature), V (Exhaust Vacuum), AP (Ambient Pressure), RH (Relative Humidity)")

# features and target
feature_cols_3 = ['AT', 'V', 'AP', 'RH']
X3 = df3[feature_cols_3].values
y3 = df3['PE'].values

# handle missing if any
missing_rows = df3.isnull().any(axis=1).sum()
print(f"\nRows with missing: {missing_rows}")
if missing_rows>0:
    df3 = df3.dropna()
    X3 = df3[feature_cols_3].values
    y3 = df3['PE'].values
    print(f"After dropping: {df3.shape}")

# 3.2 split
X3_train, X3_test, y3_train, y3_test = train_test_split(X3, y3, test_size=0.3, random_state=42)
print(f"\nTraining: {len(X3_train)}")
print(f"Testing: {len(X3_test)}")

# 3.3 train models

lr3 = LinearRegression()
lr3.fit(X3_train, y3_train)

ridge3 = Ridge(alpha=1.0)
ridge3.fit(X3_train, y3_train)

lasso3 = Lasso(alpha=1.0)
lasso3.fit(X3_train, y3_train)

poly2_3 = PolynomialFeatures(degree=2)
X3_train_poly2 = poly2_3.fit_transform(X3_train)
X3_test_poly2 = poly2_3.transform(X3_test)

poly3_3 = PolynomialFeatures(degree=3)
X3_train_poly3 = poly3_3.fit_transform(X3_train)
X3_test_poly3 = poly3_3.transform(X3_test)

lr_poly2_3 = LinearRegression()
lr_poly2_3.fit(X3_train_poly2, y3_train)

lr_poly3_3 = LinearRegression()
lr_poly3_3.fit(X3_train_poly3, y3_train)

ridge_poly2_3 = Ridge(alpha=1.0)
ridge_poly2_3.fit(X3_train_poly2, y3_train)

ridge_poly3_3 = Ridge(alpha=1.0)
ridge_poly3_3.fit(X3_train_poly3, y3_train)

lasso_poly2_3 = Lasso(alpha=1.0)
lasso_poly2_3.fit(X3_train_poly2, y3_train)

lasso_poly3_3 = Lasso(alpha=1.0)
lasso_poly3_3.fit(X3_train_poly3, y3_train)

print(f"Models trained, alpha = 1.0")

# 3.4 evaluation
res3 = []

y3_train_pred=lr3.predict(X3_train)
y3_test_pred=lr3.predict(X3_test)
res3.append(['Linear Regression', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=ridge3.predict(X3_train)
y3_test_pred=ridge3.predict(X3_test)
res3.append(['Ridge Regression', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=lasso3.predict(X3_train)
y3_test_pred=lasso3.predict(X3_test)
res3.append(['Lasso Regression', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=lr_poly2_3.predict(X3_train_poly2)
y3_test_pred=lr_poly2_3.predict(X3_test_poly2)
res3.append(['Poly2 + Linear', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=lr_poly3_3.predict(X3_train_poly3)
y3_test_pred=lr_poly3_3.predict(X3_test_poly3)
res3.append(['Poly3 + Linear', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=ridge_poly2_3.predict(X3_train_poly2)
y3_test_pred=ridge_poly2_3.predict(X3_test_poly2)
res3.append(['Poly2 + Ridge', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=ridge_poly3_3.predict(X3_train_poly3)
y3_test_pred=ridge_poly3_3.predict(X3_test_poly3)
res3.append(['Poly3 + Ridge', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=lasso_poly2_3.predict(X3_train_poly2)
y3_test_pred=lasso_poly2_3.predict(X3_test_poly2)
res3.append(['Poly2 + Lasso', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

y3_train_pred=lasso_poly3_3.predict(X3_train_poly3)
y3_test_pred=lasso_poly3_3.predict(X3_test_poly3)
res3.append(['Poly3 + Lasso', mean_squared_error(y3_train, y3_train_pred), mean_absolute_error(y3_train, y3_train_pred), r2_score(y3_train, y3_train_pred), mean_squared_error(y3_test, y3_test_pred), mean_absolute_error(y3_test, y3_test_pred), r2_score(y3_test, y3_test_pred)])

df_res3 = pd.DataFrame(res3, columns=['Model', 'Train MSE', 'Train MAE', 'Train R2', 'Test MSE', 'Test MAE', 'Test R2'])
print(df_res3.to_string(index=False))

# 3.5 find best
best_idx=0
best_r2=res3[0][6]
i=1
while i<len(res3):
    if res3[i][6]>best_r2:
        best_r2=res3[i][6]
        best_idx=i
    i+=1
print(f"Best model: {res3[best_idx][0]} with test R2={best_r2:.4f}")





Part 3: Power Plant Data -> PE Prediction

3.1 Load and Inspect Dataset
Column names: ['AT', 'V', 'AP', 'RH', 'PE']
Shape: (9568, 5)

Summary Statistics:
                AT            V           AP           RH           PE
count  9568.000000  9568.000000  9568.000000  9568.000000  9568.000000
mean     19.651231    54.305804  1013.259078    73.308978   454.365009
std       7.452473    12.707893     5.938784    14.600269    17.066995
min       1.810000    25.360000   992.890000    25.560000   420.260000
25%      13.510000    41.740000  1009.100000    63.327500   439.750000
50%      20.345000    52.080000  1012.940000    74.975000   451.550000
75%      25.720000    66.540000  1017.260000    84.830000   468.430000
max      37.110000    81.560000  1033.300000   100.160000   495.760000

Missing values:
AT    0
V     0
AP    0
RH    0
PE    0
dtype: int64

Target (DV): PE (Electrical Energy Output)
Features (IVs): AT (Ambient Temperature), V (Exhaust Vacuum), AP (Ambient Pressure), RH (R

/home/veronica/桌面/Homework1/.venv/lib/python3.12/site-packages/sklearn/linear_model/_ridge.py:228: LinAlgWarning: An ill-conditioned matrix detected: slice 0 has rcond = 2.598279743128814e-19.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
/home/veronica/桌面/Homework1/.venv/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.336e+04, tolerance: 1.945e+02
  model = cd_fast.enet_coordinate_descent(


Models trained, alpha = 1.0
            Model  Train MSE  Train MAE  Train R2  Test MSE  Test MAE  Test R2
Linear Regression  20.580840   3.607182  0.929136 21.239857  3.649933 0.927548
 Ridge Regression  20.580840   3.607186  0.929136 21.239805  3.649931 0.927549
 Lasso Regression  20.647117   3.617731  0.928908 21.247793  3.649382 0.927521
   Poly2 + Linear  17.887800   3.333026  0.938409 18.647312  3.386145 0.936392
   Poly3 + Linear  17.079895   3.211027  0.941191 17.614004  3.264772 0.939917
    Poly2 + Ridge  17.894772   3.333950  0.938385 18.643907  3.386515 0.936404
    Poly3 + Ridge  17.136922   3.220419  0.940994 17.699075  3.274420 0.939626
    Poly2 + Lasso  18.507546   3.411508  0.936275 19.215444  3.460524 0.934454
    Poly3 + Lasso  18.260967   3.360199  0.937124 18.918335  3.409587 0.935467
Best model: Poly3 + Linear with test R2=0.9399


/home/veronica/桌面/Homework1/.venv/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 6.145e+04, tolerance: 1.945e+02
  model = cd_fast.enet_coordinate_descent(


Poly3 + Linear has the best test R2 = 0.9399. This is higher than basic Linear Regression which has test R2 = 0.9275.

The polynomial models (especially degree 3) perform better than linear models. This suggests that PE (electrical energy output) has a nonlinear relationship with the input features (AT, V, AP, RH).

Ridge Regression has almost the same performance as Linear Regression (both around 0.9275). Lasso is a little bit worse. So regularization does not help much on this dataset. I think this is because the features are not very correlated with each other, so there is no need for regularization to reduce overfitting.

The best model is Poly3 + Linear because polynomial features can capture the nonlinear patterns in power plant data.